pkg/portworx/k8s.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright © 2019 Portworx
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package portworx
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/portworx/px/pkg/util"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kclikube "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type KubeConnectionData struct {
+	ClientConfig clientcmd.ClientConfig
+	ClientSet    *kclikube.Clientset
+}
+
+func NewCOps(kc *KubeConnectionData) COps {
+	return kc
+}
+
+func (p *KubeConnectionData) GetPodsByLabels(
+	namespace string,
+	labels string,
+) ([]v1.Pod, error) {
+	if p.ClientSet == nil {
+		return make([]v1.Pod, 0), nil
+	}
+	podClient := p.ClientSet.CoreV1().Pods(namespace)
+	lo := metav1.ListOptions{}
+	if len(labels) > 0 {
+		lo.LabelSelector = labels
+	}
+	podList, err := podClient.List(lo)
+	if err != nil {
+		return nil, err
+	}
+	return podList.Items, nil
+}
+
+func (p *KubeConnectionData) GetPvcsByLabels(
+	namespace string,
+	labels string,
+) ([]v1.PersistentVolumeClaim, error) {
+	if p.ClientSet == nil {
+		return make([]v1.PersistentVolumeClaim, 0), nil
+	}
+	p.ClientSet.CoreV1().Pods(namespace)
+	pvcClient := p.ClientSet.CoreV1().PersistentVolumeClaims(namespace)
+
+	lo := metav1.ListOptions{}
+	if len(labels) > 0 {
+		lo.LabelSelector = labels
+	}
+	pvcList, err := pvcClient.List(lo)
+	if err != nil {
+		return nil, err
+	}
+
+	return pvcList.Items, nil
+}
+
+type logPayload struct {
+	rw           rest.ResponseWrapper
+	podName      string
+	podNamespace string
+}
+
+func (p *KubeConnectionData) GetLogs(
```

Comment:
Somehow (don't know yet) we may need a unit test for these functions. This may be a task for Milestone 2.

Reply:
Yes we need to have a way to simulate a k8s  setup in UTs